src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/NuGetPowerShellUsageCollector.cs

```diff
@@ -0,0 +1,371 @@
+        {
+            lock (_lock)
+            {
+                // Edge case: PMC used without solution load
+                if (!_vsSolutionData.SolutionLoaded && _vsSolutionData.PMCExecuteCommandCount > 0)
+                {
+                    // PMC used before any solution is loaded, let's emit what we have for nugetvsinstanceclose event.
+                    TelemetryActivity.EmitTelemetryEvent(_vsSolutionData.ToTelemetryEvent());
+                }
+
+                // Emit VS Instance telemetry
+                TelemetryActivity.EmitTelemetryEvent(_vsInstanceData.ToTelemetryEvent());
+            }
+        }
+
+        //If open new solution then need to clear previous solution events. But powershell remain loaded in memory.
+        private void ClearSolutionData()
+        {
+            bool pmcPowershellLoad = _vsSolutionData.LoadedFromPMC;
+            bool pmuiPowershellLoad = _vsSolutionData.LoadedFromPMUI;
+            int pmcWindowLoadCount = _vsSolutionData.PMCWindowLoadCount;
+
+            _vsSolutionData = new SolutionData();
+
+            _vsSolutionData.LoadedFromPMC = pmcPowershellLoad;
+            _vsSolutionData.LoadedFromPMUI = pmuiPowershellLoad;
+            _vsSolutionData.PMCWindowLoadCount = pmcWindowLoadCount;
+        }
+
+        public void Dispose()
+        {
+            NuGetPowerShellUsage.PowerShellLoadEvent -= NuGetPowerShellUsage_PMCLoadEventHandler;
+            NuGetPowerShellUsage.PowerShellCommandExecuteEvent -= NuGetPowerShellUsage_PowerShellCommandExecuteEventHandler;
+            NuGetPowerShellUsage.NuGetCmdletExecutedEvent -= NuGetPowerShellUsage_NuGetCmdletExecutedEventHandler;
+            NuGetPowerShellUsage.InitPs1LoadEvent -= NuGetPowerShellUsage_InitPs1LoadEventHandler;
+            NuGetPowerShellUsage.PMCWindowsEvent -= NuGetPowerShellUsage_PMCWindowsEventHandler;
+            NuGetPowerShellUsage.SolutionOpenEvent -= NuGetPowerShellUsage_SolutionOpenHandler;
+            NuGetPowerShellUsage.SolutionCloseEvent -= NuGetPowerShellUsage_SolutionCloseHandler;
+            NuGetPowerShellUsage.VSInstanceCloseEvent -= NuGetPowerShellUsage_VSInstanseCloseHandler;
+        }
+
+        internal class SolutionData
+        {
+            internal bool FirstTimeLoadedFromPMC { get; set; }
+            internal bool FirstTimeLoadedFromPMUI { get; set; }
+            internal bool InitPs1LoadedFromPMCFirst { get; set; }
+            internal bool InitPs1LoadPMC { get; set; }
+            internal bool InitPs1LoadPMUI { get; set; }
+            internal bool LoadedFromPMC { get; set; }
+            internal bool LoadedFromPMUI { get; set; }
+            internal bool NuGetCommandUsed { get; set; }
+            internal int PMCExecuteCommandCount { get; set; }
+            internal int PMCWindowLoadCount { get; set; }
+            internal int PMUIExecuteCommandCount { get; set; }
+            internal bool SolutionLoaded { get; set; }
```

Comment:
How can `SolutionLoaded == false` when this class is specifically for solution session data?